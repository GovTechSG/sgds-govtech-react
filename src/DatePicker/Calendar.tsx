import * as React from 'react';
interface CalendarProps extends React.HTMLAttributes<HTMLTableElement> {
  selectedDate: Date[];
  displayDate: Date;
  minDate?: string;
  maxDate?: string;
  changeDate: (date: Date) => void;
  dayLabels?: string[];
  mode: 'single' | 'range';
}
const CELL_PADDING = '5px';
const DAY_LABELS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
const daysInMonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

const setTimeToNoon = (date: Date) => {
  const newDate = new Date(date);
  newDate.setHours(12);
  newDate.setMinutes(0);
  newDate.setSeconds(0);
  newDate.setMilliseconds(0);
  return newDate;
};
const generateIncrementDays = (start: Date, end: Date) => {
  let arr = [];
  for (let dt = start; dt <= end; dt.setDate(dt.getDate() + 1)) {
    arr.push(new Date(dt));
  }
  return arr;
};
export const Calendar = React.forwardRef<HTMLTableElement, CalendarProps>(
  ({ dayLabels = DAY_LABELS, ...props }, ref) => {
    const handleClick = (e: React.MouseEvent<HTMLTableCellElement>) => {
      const day = e.currentTarget.getAttribute('data-day')!;
      const displayDateClone = new Date(props.displayDate);
      const newSelectedDate = setTimeToNoon(displayDateClone);
      newSelectedDate.setDate(parseInt(day));
      props.changeDate(newSelectedDate);
    };
    const currentDate = setTimeToNoon(new Date());
    const selectedDates = props.selectedDate.map((d) => setTimeToNoon(d));
    const rangeSelectedDates = generateIncrementDays(
      new Date(selectedDates[0]),
      new Date(selectedDates[1])
    );
    const minimumDate = props.minDate
      ? setTimeToNoon(new Date(props.minDate))
      : null;
    const maximumDate = props.maxDate
      ? setTimeToNoon(new Date(props.maxDate))
      : null;
    const year = props.displayDate.getFullYear();
    const month = props.displayDate.getMonth();
    const firstDay = new Date(year, month, 1);
    const startingDay = firstDay.getDay();
    let monthLength = daysInMonth[month];
    if (month == 1) {
      if ((year % 4 == 0 && year % 100 != 0) || year % 400 == 0) {
        monthLength = 29;
      }
    }

    const weeks = [];
    let day = 1;
    for (let i = 0; i < 9; i++) {
      const week = [];
      for (let j = 0; j <= 6; j++) {
        if (day <= monthLength && (i > 0 || j >= startingDay)) {
          let className = undefined;
          const date = new Date(year, month, day, 12, 0, 0, 0).toISOString();
          const beforeMinDate =
            minimumDate &&
            Date.parse(date) < Date.parse(minimumDate.toISOString());
          const afterMinDate =
            maximumDate &&
            Date.parse(date) > Date.parse(maximumDate.toISOString());

          let clickHandler: React.MouseEventHandler | undefined = handleClick;
          const style = {
            cursor: 'pointer',
            padding: CELL_PADDING,
            borderRadius: 50,
          };

          if (beforeMinDate || afterMinDate) {
            className = 'text-muted';
            clickHandler = undefined;
            style.cursor = 'default';
          }

          if (Date.parse(date) === Date.parse(currentDate.toISOString())) {
            //if selected Date is not current Date
            className = 'text-primary';
          }
          if (selectedDates.length > 0) {
            rangeSelectedDates.forEach((d) => {
              if (Date.parse(date) === Date.parse(d.toISOString())) {
                className = 'bg-primary';
              }
            });
            if (
              Date.parse(date) === Date.parse(selectedDates[0]!.toISOString())
            ) {
              className = 'bg-primary';
            }
          }

          week.push(
            <td
              key={j}
              data-day={day}
              onClick={clickHandler}
              style={style}
              className={className}
            >
              {day}
            </td>
          );
          day++;
        } else {
          week.push(<td key={j} />);
        }
      }

      weeks.push(<tr key={i}>{week}</tr>);
      if (day > monthLength) {
        break;
      }
    }

    return (
      <table className="text-center" ref={ref}>
        <thead>
          <tr>
            {dayLabels.map((label: string, index: number) => {
              return (
                <td
                  key={index}
                  className="text-muted"
                  style={{ padding: CELL_PADDING }}
                >
                  <small>{label}</small>
                </td>
              );
            })}
          </tr>
        </thead>
        <tbody>{weeks}</tbody>
      </table>
    );
  }
);

export default Calendar;
