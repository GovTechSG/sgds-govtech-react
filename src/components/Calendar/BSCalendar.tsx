import * as React from 'react';
import Button from '../Button/Button';
interface CalendarProps extends React.HTMLAttributes<HTMLTableElement> {
  selectedDate: Date;
  displayDate: Date;
  minDate?: string;
  maxDate?: string;
  changeDate: (date: Date) => void;
  dayLabels?: string[];
  cellPadding: string;
  weekStartsOn: number;
  showTodayButton: boolean;
  todayButtonLabel: string;
  roundedCorners: boolean;
  showWeeks: boolean;
}
const MAX_DATE = '2021-05-19T12:00:00.000Z';
const MIN_DATE = '2020-05-19T12:00:00.000Z';
const DAY_LABELS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
const defaultProps = {
  dayLabels : ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
  cellPadding: '5px',
};
const daysInMonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

const setTimeToNoon = (date: Date) => {
  date.setHours(12);
  date.setMinutes(0);
  date.setSeconds(0);
  date.setMilliseconds(0);
  return date;
};
const getWeekNumber = (date: Date) => {
  const target = new Date(date.valueOf());
  const dayNr = (date.getDay() + 6) % 7;
  target.setDate(target.getDate() - dayNr + 3);
  const firstThursday = target.valueOf();
  target.setMonth(0, 1);
  if (target.getDay() !== 4) {
    target.setMonth(0, 1 + ((4 - target.getDay() + 7) % 7));
  }
  return 1 + Math.ceil((firstThursday - target.valueOf()) / 604800000);
};
export const Calendar = React.forwardRef<HTMLTableElement, CalendarProps>(
  (
    {
      dayLabels = DAY_LABELS,
      cellPadding = '5px',
      showWeeks = false,
      todayButtonLabel = 'Today',
      roundedCorners = false,
      ...props
    },
    ref
  ) => {
    const handleClick = (e: React.MouseEvent<HTMLTableCellElement>) => {
      const day = e.currentTarget.getAttribute('data-day')!;
      const newSelectedDate = setTimeToNoon(props.displayDate!);
      newSelectedDate.setDate(parseInt(day));
      props.changeDate(newSelectedDate);
    };
    const handleClickToday = () => {
      const newSelectedDate = setTimeToNoon(new Date());
      props.changeDate(newSelectedDate);
    };
  
    const currentDate = setTimeToNoon(new Date());
    const selectedDate = setTimeToNoon(props.selectedDate);
    const minimumDate = props.minDate
      ? setTimeToNoon(new Date(props.minDate))
      : null;
    const maximumDate = props.maxDate
      ? setTimeToNoon(new Date(props.maxDate))
      : null;
    const year = props.displayDate.getFullYear();
    const month = props.displayDate.getMonth();
    const firstDay = new Date(year, month, 1);
    const startingDay =
      props.weekStartsOn > 1
        ? firstDay.getDay() - props.weekStartsOn + 7
        : props.weekStartsOn === 1
        ? firstDay.getDay() === 0
          ? 6
          : firstDay.getDay() - 1
        : firstDay.getDay();
  
    let monthLength = daysInMonth[month];
    if (month == 1) {
      if ((year % 4 == 0 && year % 100 != 0) || year % 400 == 0) {
        monthLength = 29;
      }
    }
  
    const weeks = [];
    let day = 1;
    for (let i = 0; i < 9; i++) {
      const week = [];
      for (let j = 0; j <= 6; j++) {
        if (day <= monthLength && (i > 0 || j >= startingDay)) {
          let className = undefined;
          const date = new Date(year, month, day, 12, 0, 0, 0).toISOString();
          const beforeMinDate =
            minimumDate &&
            Date.parse(date) < Date.parse(minimumDate.toISOString());
          const afterMinDate =
            maximumDate &&
            Date.parse(date) > Date.parse(maximumDate.toISOString());
  
          let clickHandler: React.MouseEventHandler | undefined = handleClick;
          const style = {
            cursor: 'pointer',
            padding: cellPadding,
            borderRadius: roundedCorners ? 5 : 0,
          };
  
          if (beforeMinDate || afterMinDate) {
            className = 'text-muted';
            clickHandler = undefined;
            style.cursor = 'default';
          } else if (
            Date.parse(date) === Date.parse(selectedDate!.toISOString())
          ) {
            className = 'bg-primary';
          } else if (Date.parse(date) === Date.parse(currentDate.toISOString())) {
            className = 'text-primary';
          }
  
          week.push(
            <td
              key={j}
              data-day={day}
              onClick={clickHandler}
              style={style}
              className={className}
            >
              {day}
            </td>
          );
          day++;
        } else {
          week.push(<td key={j} />);
        }
      }
  
      if (showWeeks) {
        const weekNum = getWeekNumber(
          new Date(year, month, day - 1, 12, 0, 0, 0)
        );
        week.unshift(
          <td
            key={7}
            style={{
              padding: cellPadding,
              fontSize: '0.8em',
              color: 'darkgrey',
            }}
            className="text-muted"
          >
            {weekNum}
          </td>
        );
      }
  
      weeks.push(<tr key={i}>{week}</tr>);
      if (day > monthLength) {
        break;
      }
    }
  
    const weekColumn = showWeeks ? (
      <td className="text-muted current-week" style={{ padding: cellPadding }} />
    ) : null;
    return (
      <table className="text-center" ref={ref} >
        <thead>
          <tr>
            {weekColumn}
            {dayLabels.map((label: string, index: number) => {
              return (
                <td
                  key={index}
                  className="text-muted"
                  style={{ padding: cellPadding }}
                >
                  <small>{label}</small>
                </td>
              );
            })}
          </tr>
        </thead>
        <tbody>{weeks}</tbody>
        {props.showTodayButton && (
          <tfoot>
            <tr>
              <td colSpan={dayLabels.length} style={{ paddingTop: '9px' }}>
                <Button
                  //   block
                  size="sm"
                  className="u-today-button"
                  onClick={handleClickToday}
                >
                  {todayButtonLabel}
                </Button>
              </td>
            </tr>
          </tfoot>
        )}
      </table>
    );
  }
);
// export const Calendar: React.FC<CalendarProps> = ({
//   dayLabels = DAY_LABELS,
//   cellPadding = '5px',
//   showWeeks = false,
//   todayButtonLabel = 'Today',
//   roundedCorners = false,
//   ...props
// }) => {
//   const handleClick = (e: React.MouseEvent<HTMLTableCellElement>) => {
//     const day = e.currentTarget.getAttribute('data-day')!;
//     const newSelectedDate = setTimeToNoon(props.displayDate!);
//     newSelectedDate.setDate(parseInt(day));
//     props.changeDate(newSelectedDate);
//   };
//   const handleClickToday = () => {
//     const newSelectedDate = setTimeToNoon(new Date());
//     props.changeDate(newSelectedDate);
//   };

//   const currentDate = setTimeToNoon(new Date());
//   const selectedDate = setTimeToNoon(props.selectedDate);
//   const minimumDate = props.minDate
//     ? setTimeToNoon(new Date(props.minDate))
//     : null;
//   const maximumDate = props.maxDate
//     ? setTimeToNoon(new Date(props.maxDate))
//     : null;
//   const year = props.displayDate.getFullYear();
//   const month = props.displayDate.getMonth();
//   const firstDay = new Date(year, month, 1);
//   const startingDay =
//     props.weekStartsOn > 1
//       ? firstDay.getDay() - props.weekStartsOn + 7
//       : props.weekStartsOn === 1
//       ? firstDay.getDay() === 0
//         ? 6
//         : firstDay.getDay() - 1
//       : firstDay.getDay();

//   let monthLength = daysInMonth[month];
//   if (month == 1) {
//     if ((year % 4 == 0 && year % 100 != 0) || year % 400 == 0) {
//       monthLength = 29;
//     }
//   }

//   const weeks = [];
//   let day = 1;
//   for (let i = 0; i < 9; i++) {
//     const week = [];
//     for (let j = 0; j <= 6; j++) {
//       if (day <= monthLength && (i > 0 || j >= startingDay)) {
//         let className = undefined;
//         const date = new Date(year, month, day, 12, 0, 0, 0).toISOString();
//         const beforeMinDate =
//           minimumDate &&
//           Date.parse(date) < Date.parse(minimumDate.toISOString());
//         const afterMinDate =
//           maximumDate &&
//           Date.parse(date) > Date.parse(maximumDate.toISOString());

//         let clickHandler: React.MouseEventHandler | undefined = handleClick;
//         const style = {
//           cursor: 'pointer',
//           padding: cellPadding,
//           borderRadius: roundedCorners ? 5 : 0,
//         };

//         if (beforeMinDate || afterMinDate) {
//           className = 'text-muted';
//           clickHandler = undefined;
//           style.cursor = 'default';
//         } else if (
//           Date.parse(date) === Date.parse(selectedDate!.toISOString())
//         ) {
//           className = 'bg-primary';
//         } else if (Date.parse(date) === Date.parse(currentDate.toISOString())) {
//           className = 'text-primary';
//         }

//         week.push(
//           <td
//             key={j}
//             data-day={day}
//             onClick={clickHandler}
//             style={style}
//             className={className}
//           >
//             {day}
//           </td>
//         );
//         day++;
//       } else {
//         week.push(<td key={j} />);
//       }
//     }

//     if (showWeeks) {
//       const weekNum = getWeekNumber(
//         new Date(year, month, day - 1, 12, 0, 0, 0)
//       );
//       week.unshift(
//         <td
//           key={7}
//           style={{
//             padding: cellPadding,
//             fontSize: '0.8em',
//             color: 'darkgrey',
//           }}
//           className="text-muted"
//         >
//           {weekNum}
//         </td>
//       );
//     }

//     weeks.push(<tr key={i}>{week}</tr>);
//     if (day > monthLength) {
//       break;
//     }
//   }

//   const weekColumn = showWeeks ? (
//     <td className="text-muted current-week" style={{ padding: cellPadding }} />
//   ) : null;
//   return (
//     <table className="text-center">
//       <thead>
//         <tr>
//           {weekColumn}
//           {dayLabels.map((label: string, index: number) => {
//             return (
//               <td
//                 key={index}
//                 className="text-muted"
//                 style={{ padding: cellPadding }}
//               >
//                 <small>{label}</small>
//               </td>
//             );
//           })}
//         </tr>
//       </thead>
//       <tbody>{weeks}</tbody>
//       {props.showTodayButton && (
//         <tfoot>
//           <tr>
//             <td colSpan={dayLabels.length} style={{ paddingTop: '9px' }}>
//               <Button
//                 //   block
//                 size="sm"
//                 className="u-today-button"
//                 onClick={handleClickToday}
//               >
//                 {todayButtonLabel}
//               </Button>
//             </td>
//           </tr>
//         </tfoot>
//       )}
//     </table>
//   );
// };

export default Calendar;
