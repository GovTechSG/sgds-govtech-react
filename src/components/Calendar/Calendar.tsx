import * as React from 'react';
import ReactCalendar, {CalendarProps as ReactCalendarProps} from 'react-calendar';

export interface CalendarProps extends ReactCalendarProps {}

export const Calendar = (props: CalendarProps) => (
     <ReactCalendar {...props}/>
)

export default Calendar