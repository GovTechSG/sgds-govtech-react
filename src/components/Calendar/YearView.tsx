import classNames from 'classnames';
import * as React from 'react';
import { useState } from 'react';
import Button from '../Button/Button';
import Col from '../Col';
export interface YearViewProps extends React.HTMLAttributes<HTMLElement> {
  displayDate: Date;
}



export const YearView = React.forwardRef<HTMLDivElement, YearViewProps>(
  ({ displayDate, ...props }, ref) => {
      console.log(displayDate.getFullYear())
      const displayYear = displayDate.getFullYear()
      const startLimit = displayYear - 5
      const endLimit = displayYear + 6
      const yearArray = []

      for(let i = startLimit; i<endLimit + 1; i++){
          yearArray.push(i)
      }
    // const [year, setYear] = useState(displayDate.getFullYear())
    // const [startLimit, setStartLimit]  = useState(displayDate.getFullYear() - 5)
    // const [endLimit, setEndLimit] = useState(displayDate.getFullYear() + 6)
    return (
      <div className="container text-center" ref={ref} {...props}>
        <div className="row">
          {yearArray.map((y) => (
            <div
              className={classNames(displayYear === y && 'active', 'col-4')}
              key={y}
            >
              <button>{y}</button>
            </div>
          ))}
        </div>
      </div>
    );
  }
);

export default YearView;
