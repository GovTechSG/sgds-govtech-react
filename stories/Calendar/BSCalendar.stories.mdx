import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import { BSCalendar } from '../../src/components/index';
import { useState } from 'react';


<Meta title="MDX/BSCalendar" component={BSCalendar} />

export const BSCalendarTemplate = () => {
  const [value, setValue] =useState(new Date().toISOString())
  const handleChange = (value, formattedValue ) => {
    setValue({value , formattedValue})
  }
    return <BSCalendar value={value} onChange={handleChange}/>
}

<Canvas>
  <Story name="Default">
      {BSCalendarTemplate.bind({})}
  </Story>
</Canvas>