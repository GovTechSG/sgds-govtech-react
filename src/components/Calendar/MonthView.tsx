import classNames from 'classnames';
import * as React from 'react';
import Button from '../Button/Button';
import Col from '../Col';
import DatePickerContext from './DatePickerContext'
import { useState, useContext } from 'react';
import { CalendarState } from './DatePicker';
export interface MonthViewProps extends React.HTMLAttributes<HTMLElement> {
  displayDate: Date;
  state: CalendarState;
  setState: React.Dispatch<React.SetStateAction<CalendarState>>
}


const MONTH_LABELS = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];

export const MonthView = React.forwardRef<HTMLDivElement, MonthViewProps>(
  ({ displayDate, state, setState, ...props }, ref) => {
    const displayMonth = MONTH_LABELS[displayDate.getMonth()];
    const {setView}  = useContext(DatePickerContext)

    const handleClickMonth = (month: number) => {
      setView('day')
      setState({
        ...state,
        displayDate: new Date(state.displayDate.setMonth(month))
      })
    }
    return (
      <div className="container text-center" ref={ref} {...props}>
        <div className="row">
          {MONTH_LABELS.map((m, idx) => (
            <div
              className={classNames(displayMonth === m && 'active', 'col-4')}
              key={m}
            >
              <button onClick={() =>handleClickMonth(idx)}>{m}</button>
            </div>
          ))}
        </div>
      </div>
    );
  }
);

export default MonthView;
