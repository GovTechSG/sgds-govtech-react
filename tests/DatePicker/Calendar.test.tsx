import * as React from 'react';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import Calendar, {
  DAY_LABELS,
  daysInMonth,
  setTimeToNoon,
  generateIncrementDays,
} from '../../src/DatePicker/Calendar';

describe('setTimeToNoon function', () => {
  it('given date prop, should return new Date hours set to 12, min 0, sec 0, ms 0', () => {
    const result = setTimeToNoon(new Date(2022, 2, 1, 13, 14, 15, 7));

    expect(result).toEqual(new Date(2022, 2, 1, 12, 0, 0, 0));
  });

  it('given date already in noon, should return new Date hours set to 12, min 0, sec 0, ms 0', () => {
    const result = setTimeToNoon(new Date(2022, 2, 1, 12, 0, 0, 0));

    expect(result).toEqual(new Date(2022, 2, 1, 12, 0, 0, 0));
  });
  it('given any date  should return new Date hours set to 12, min 0, sec 0, ms 0', () => {
    const result = setTimeToNoon(new Date());
    const assertion = new Date();
    assertion.setHours(12);
    assertion.setMinutes(0);
    assertion.setSeconds(0);
    assertion.setMilliseconds(0);
    expect(result).toEqual(assertion);
  });
});

describe('generateIncrementDays fn', () => {
  it('given start Date and end Date set to noon, returns array of the consecutively incrementing date from start till end', () => {
    const start = new Date(2020, 0, 1, 12, 0, 0, 0);
    const end = new Date(2020, 0, 10, 12, 0, 0, 0);
    const result = generateIncrementDays(start, end);

    expect(result.length).toEqual(10);
    expect(result[0]).toEqual(new Date(2020, 0, 1, 12, 0, 0, 0));
    expect(result[result.length - 1]).toEqual(
      new Date(2020, 0, 10, 12, 0, 0, 0)
    );
  });
  it('given start Date and end Date set to noon as same value, returns array of the consecutively incrementing date from start till end', () => {
    const start = new Date(2020, 0, 1, 12, 0, 0, 0);
    const end = new Date(2020, 0, 1, 12, 0, 0, 0);
    const result = generateIncrementDays(start, end);

    expect(result.length).toEqual(1);
    expect(result[0]).toEqual(new Date(2020, 0, 1, 12, 0, 0, 0));
    expect(result[result.length - 1]).toEqual(
      new Date(2020, 0, 1, 12, 0, 0, 0)
    );
  });
  it('given start Date and end Date in diff months, returns correct array of dates', () => {
    const start = new Date(2019, 11, 31, 12, 0, 0, 0);
    const end = new Date(2020, 0, 2, 12, 0, 0, 0);
    const result = generateIncrementDays(start, end);

    expect(result.length).toEqual(3);
    expect(result[0]).toEqual(new Date(2019, 11, 31, 12, 0, 0, 0));
    expect(result[1]).toEqual(new Date(2020, 0, 1, 12, 0, 0, 0));
    expect(result[2]).toEqual(new Date(2020, 0, 2, 12, 0, 0, 0));
  });
  it('given start Date later than end Date, fn should rearrange to return array in ascending order', () => {
    const end = new Date(2019, 11, 31, 12, 0, 0, 0);
    const start = new Date(2020, 0, 2, 12, 0, 0, 0);
    const result = generateIncrementDays(start, end);

    expect(result.length).toEqual(3);
    expect(result[0]).toEqual(new Date(2019, 11, 31, 12, 0, 0, 0));
    expect(result[1]).toEqual(new Date(2020, 0, 1, 12, 0, 0, 0));
    expect(result[2]).toEqual(new Date(2020, 0, 2, 12, 0, 0, 0));
  });
});

describe('Single mode Calendar', () => {
  const mode = 'single';
  it('should have the default html structure', () => {
    const displayDate = new Date(2022, 2, 21);
    const selectedDate = [new Date(2022, 2, 18)];
    const mockChangeDate = jest.fn();
    const { container, asFragment, getByText } = render(
      <Calendar
        changeDate={mockChangeDate}
        mode={mode}
        displayDate={displayDate}
        selectedDate={selectedDate}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(container.querySelector('table.text-center')).toBeInTheDocument();
    DAY_LABELS.forEach((day) => {
      expect(getByText(day)).toBeInTheDocument();
    });

    const totalDaysInDisplayDate = daysInMonth[displayDate.getMonth()];
    expect(getByText(totalDaysInDisplayDate)).toBeInTheDocument();
    expect(getByText(selectedDate[0].getDate()).classList).toContain(
      'bg-primary'
    );
    expect(container.querySelectorAll('.bg-primary').length).toEqual(1);
  });
  it('day should have class text-primary when its current date', () => {
    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));
    const displayDate = new Date('2020-01-23');
    const selectedDate = [new Date('2020-01-23')];
    const mockChangeDate = jest.fn();
    const { getByText, container } = render(
      <Calendar
        changeDate={mockChangeDate}
        mode={mode}
        displayDate={displayDate}
        selectedDate={selectedDate}
      />
    );

    expect(getByText('1')).toBeInTheDocument();
    expect(getByText('1').classList).toContain('text-primary');
    expect(container.querySelectorAll('.text-primary').length).toEqual(1);
    jest.clearAllTimers();
  });
  it('should show appropriate days depending on month', () => {
    const arrayOfDates = [
      new Date(2022, 0, 20),
      new Date(2022, 1, 20),
      new Date(2022, 2, 20),
      new Date(2022, 3, 20),
      new Date(2022, 4, 20),
      new Date(2022, 5, 20),
      new Date(2022, 6, 20),
      new Date(2022, 7, 20),
      new Date(2022, 8, 20),
      new Date(2022, 9, 20),
      new Date(2022, 10, 20),
      new Date(2022, 11, 20),
    ];
    const mockChangeDate = jest.fn();

    arrayOfDates.forEach((date, idx) => {
      const { getByText } = render(
        <Calendar
          changeDate={mockChangeDate}
          mode={mode}
          displayDate={date}
          selectedDate={[date]}
        />
      );
      const expectedLastDatePerMonth = daysInMonth[idx];
      expect(getByText(expectedLastDatePerMonth)).toBeInTheDocument();
      cleanup();
    });
  });
  it('account for leap year in feb months', () => {
    const mockChangeDate = jest.fn();
    const leapYear = [2000, 2020, 2024, 2028];
    leapYear.forEach((y) => {
      const feb = new Date(y, 1, 20);
      const { getByText } = render(
        <Calendar
          changeDate={mockChangeDate}
          mode={mode}
          displayDate={feb}
          selectedDate={[feb]}
        />
      );
      expect(getByText('29')).toBeInTheDocument();
      cleanup();
    });
  });
  it('end of century years but not leap year', () => {
    const mockChangeDate = jest.fn();
    const notLeapYear = [1700, 1800, 1900, 2100];
    notLeapYear.forEach((y) => {
      const feb = new Date(y, 1, 20);
      const { queryByText } = render(
        <Calendar
          changeDate={mockChangeDate}
          mode={mode}
          displayDate={feb}
          selectedDate={[feb]}
        />
      );
      expect(queryByText('29')).not.toBeInTheDocument();
      cleanup();
    });
  });
  it('bg-primary changes with selectedDate ', () => {
    const mockChangeDate = jest.fn();
    const date = new Date(2022, 2, 20);
    const { getByText, rerender } = render(
      <Calendar
        changeDate={mockChangeDate}
        mode={mode}
        displayDate={date}
        selectedDate={[date]}
      />
    );
    expect(getByText('20').classList).toContain('bg-primary');
    rerender(
      <Calendar
        changeDate={mockChangeDate}
        mode={mode}
        displayDate={date}
        selectedDate={[new Date(2022, 2, 1)]}
      />
    );

    expect(getByText('20').classList).not.toContain('bg-primary');
    expect(getByText('1').classList).toContain('bg-primary');
  });
  it('clickhandler is fired when click of dates', async () => {
    const mockChangeDate = jest.fn();
    const date = new Date(2022, 2, 20);
    const { getByText } = render(
      <Calendar
        changeDate={mockChangeDate}
        mode={mode}
        displayDate={date}
        selectedDate={[date]}
      />
    );
    fireEvent.click(getByText('21'));
    await waitFor(() => {
      expect(mockChangeDate).toHaveBeenCalled();
    });
    fireEvent.click(getByText('21'));
    await waitFor(() => {
      expect(mockChangeDate).toHaveBeenCalled();
    });
  });
  it('clickhandler not fired when click of dates out of min and max range', async () => {
    const mockChangeDate = jest.fn();
    const date = new Date(2022, 2, 20);
    const minDate = new Date(2022, 2, 15).toISOString();
    const maxDate = new Date(2022, 2, 21).toISOString();
    const { getByText } = render(
      <Calendar
        changeDate={mockChangeDate}
        mode={mode}
        displayDate={date}
        selectedDate={[date]}
        minDate={minDate}
        maxDate={maxDate}
      />
    );
    const activeRange = [15, 16, 17, 18, 19, 20, 21];
    activeRange.forEach((day) => {
      expect(getByText(day).classList).not.toContain('text-muted');
      expect(getByText(day).style.cursor).toEqual('pointer');
    });

    fireEvent.click(getByText('21'));
    await waitFor(() => {
      expect(mockChangeDate).toHaveBeenCalled();
      mockChangeDate.mockReset();
    });

    fireEvent.click(getByText('15'));
    await waitFor(() => {
      expect(mockChangeDate).toHaveBeenCalled();
      mockChangeDate.mockReset();
    });
    fireEvent.click(getByText('14'));
    expect(getByText('14').classList).toContain('text-muted');
    expect(getByText('14').style.cursor).toEqual('default');
    await waitFor(() => {
      expect(mockChangeDate).not.toHaveBeenCalled();
      mockChangeDate.mockReset();
    });
    fireEvent.click(getByText('22'));
    expect(getByText('22').classList).toContain('text-muted');
    expect(getByText('22').style.cursor).toEqual('default');
    await waitFor(() => {
      expect(mockChangeDate).not.toHaveBeenCalled();
      mockChangeDate.mockReset();
    });
  });
});
