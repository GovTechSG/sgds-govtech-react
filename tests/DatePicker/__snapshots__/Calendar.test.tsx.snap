// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single mode Calendar should have the default html structure 1`] = `
<DocumentFragment>
  <table
    class="text-center"
  >
    <thead>
      <tr>
        <td
          style="padding: 5px;"
        >
          <small>
            Sun
          </small>
        </td>
        <td
          style="padding: 5px;"
        >
          <small>
            Mon
          </small>
        </td>
        <td
          style="padding: 5px;"
        >
          <small>
            Tue
          </small>
        </td>
        <td
          style="padding: 5px;"
        >
          <small>
            Wed
          </small>
        </td>
        <td
          style="padding: 5px;"
        >
          <small>
            Thu
          </small>
        </td>
        <td
          style="padding: 5px;"
        >
          <small>
            Fri
          </small>
        </td>
        <td
          style="padding: 5px;"
        >
          <small>
            Sat
          </small>
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td />
        <td />
        <td
          data-day="1"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          1
        </td>
        <td
          data-day="2"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          2
        </td>
        <td
          data-day="3"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          3
        </td>
        <td
          data-day="4"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          4
        </td>
        <td
          data-day="5"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          5
        </td>
      </tr>
      <tr>
        <td
          data-day="6"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          6
        </td>
        <td
          data-day="7"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          7
        </td>
        <td
          data-day="8"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          8
        </td>
        <td
          data-day="9"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          9
        </td>
        <td
          data-day="10"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          10
        </td>
        <td
          data-day="11"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          11
        </td>
        <td
          data-day="12"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          12
        </td>
      </tr>
      <tr>
        <td
          data-day="13"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          13
        </td>
        <td
          data-day="14"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          14
        </td>
        <td
          data-day="15"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          15
        </td>
        <td
          data-day="16"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          16
        </td>
        <td
          data-day="17"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          17
        </td>
        <td
          class="bg-primary"
          data-day="18"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          18
        </td>
        <td
          data-day="19"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          19
        </td>
      </tr>
      <tr>
        <td
          data-day="20"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          20
        </td>
        <td
          data-day="21"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          21
        </td>
        <td
          class="text-primary"
          data-day="22"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          22
        </td>
        <td
          data-day="23"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          23
        </td>
        <td
          data-day="24"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          24
        </td>
        <td
          data-day="25"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          25
        </td>
        <td
          data-day="26"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          26
        </td>
      </tr>
      <tr>
        <td
          data-day="27"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          27
        </td>
        <td
          data-day="28"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          28
        </td>
        <td
          data-day="29"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          29
        </td>
        <td
          data-day="30"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          30
        </td>
        <td
          data-day="31"
          style="cursor: pointer; padding: 5px; border-radius: 50px;"
        >
          31
        </td>
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
